In [1]:
import pandas as pd

In [2]:
beginv = pd.read_csv("BegInv.csv")
endinv = pd.read_csv("EndInv.csv")
invpur = pd.read_csv("InvoicePurchases.csv")
purchases = pd.read_csv("Purchases.csv")
sales = pd.read_csv("Sales.csv")

In [3]:
beginv.rename(columns={"onHand": "Beginning Quantity"}, inplace = True)
beginv["Beginning Total $"] = (beginv["Beginning Quantity"] * beginv["Price"])
beginv.head()

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_60      1  HARDERSFIELD     60       Canadian Club 1858 VAP   
2  1_HARDERSFIELD_62      1  HARDERSFIELD     62     Herradura Silver Tequila   
3  1_HARDERSFIELD_63      1  HARDERSFIELD     63   Herradura Reposado Tequila   
4  1_HARDERSFIELD_72      1  HARDERSFIELD     72         No. 3 London Dry Gin   

    Size  Beginning Quantity  Price   startDate  Beginning Total $  
0  750mL                   8  12.99  2016-01-01             103.92  
1  750mL                   7  10.99  2016-01-01              76.93  
2  750mL                   6  36.99  2016-01-01             221.94  
3  750mL                   3  38.99  2016-01-01             116.97  
4  750mL                   6  34.99  2016-01-01             209.94

In [4]:
endinv.rename(columns={"onHand": "Ending Quantity"}, inplace = True)
endinv["Ending Total $"] = (endinv["Ending Quantity"] * endinv["Price"])
endinv.head()

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_62      1  HARDERSFIELD     62     Herradura Silver Tequila   
2  1_HARDERSFIELD_63      1  HARDERSFIELD     63   Herradura Reposado Tequila   
3  1_HARDERSFIELD_72      1  HARDERSFIELD     72         No. 3 London Dry Gin   
4  1_HARDERSFIELD_75      1  HARDERSFIELD     75    Three Olives Tomato Vodka   

    Size  Ending Quantity  Price     endDate  Ending Total $  
0  750mL               11  12.99  2016-12-31          142.89  
1  750mL                7  36.99  2016-12-31          258.93  
2  750mL                7  38.99  2016-12-31          272.93  
3  750mL                4  34.99  2016-12-31          139.96  
4  750mL                7  14.99  2016-12-31          104.93

In [5]:
invpur.head()

VendorNumber                   VendorName InvoiceDate  PONumber  \
0           105  ALTAMAR BRANDS LLC           2016-01-04      8124   
1          4466  AMERICAN VINTAGE BEVERAGE    2016-01-07      8137   
2           388  ATLANTIC IMPORTING COMPANY   2016-01-09      8169   
3           480  BACARDI USA INC              2016-01-12      8106   
4           516  BANFI PRODUCTS CORP          2016-01-07      8170   

       PODate     PayDate  Quantity    Dollars  Freight Approval  
0  2015-12-21  2016-02-16         6     214.26     3.47      NaN  
1  2015-12-22  2016-02-21        15     140.55     8.57      NaN  
2  2015-12-24  2016-02-16         5     106.60     4.61      NaN  
3  2015-12-20  2016-02-05     10100  137483.78  2935.20      NaN  
4  2015-12-24  2016-02-12      1935   15527.25   429.20      NaN

## Deliverable 1

### 1a

In [6]:
begstore = beginv.groupby(["Store"])[["Beginning Quantity", "Beginning Total $"]].sum()
begstore = begstore.sort_values(by = "Beginning Total $", ascending = False)
begstore.head(10)

Beginning Quantity  Beginning Total $
Store                                       
34                 153852         3291170.24
73                 162551         3142497.36
67                 158996         3079578.63
66                 149314         2973033.90
76                 140208         2952418.44
69                 144255         2946726.65
38                 114368         2232698.77
55                 119641         2001263.66
50                  94720         1649808.22
79                  95330         1503149.48

### 1b

In [7]:
begbrand = beginv.groupby(["Brand"])[["Beginning Quantity", "Beginning Total $"]].sum()
begbrand = begbrand.sort_values(by = "Beginning Total $", ascending = False)
begbrand.head(10)

Beginning Quantity  Beginning Total $
Brand                                       
3545                14499          463823.01
1233                12016          432455.84
8068                15341          383371.59
4261                15499          340823.01
3858                13649          327439.51
2753                 4625          286703.75
8082                 9287          278517.13
8680                 7066          275503.34
2589                 6766          270572.34
3876                14829          266773.71

## Deliverable 2

### 2a

In [8]:
endstore = endinv.groupby(["Store"])[["Ending Quantity", "Ending Total $"]].sum()
endstore = endstore.sort_values(by = "Ending Total $", ascending = False)
endstore.head(10)

Ending Quantity  Ending Total $
Store                                 
50              260717      4887260.68
73              164589      3254662.81
67              163765      3076114.82
34              145829      3074616.75
76              143866      2975945.18
69              150848      2968678.82
66              144579      2860504.99
74              166015      2803645.13
38              129397      2463906.85
55              125584      2234836.35

### 2b

In [9]:
endbrand = endinv.groupby(["Brand"])[["Ending Quantity", "Ending Total $"]].sum()
endbrand = endbrand.sort_values(by = "Ending Total $", ascending = False)
endbrand.head(10)

Ending Quantity  Ending Total $
Brand                                 
1233             15047       526494.53
3545             16770       502932.30
2753              7849       470861.51
8068             15608       366631.92
3405             12268       355649.32
4261             16769       351981.31
2757             11603       336370.97
2589              7922       300956.78
1376             13180       276648.20
2585             10487       272557.13

## Deliverable 3

In [10]:
beg_end_inv = pd.merge(left = beginv, right = endinv, how = "outer", on = ["InventoryId"])
beg_end_inv = beg_end_inv[["InventoryId", "Beginning Quantity", "Beginning Total $", "Ending Quantity", "Ending Total $"]]
beg_end_inv = beg_end_inv.fillna(0)
beg_end_inv.head()

InventoryId  Beginning Quantity  Beginning Total $  Ending Quantity  \
0  1_HARDERSFIELD_58                 8.0             103.92             11.0   
1  1_HARDERSFIELD_60                 7.0              76.93              0.0   
2  1_HARDERSFIELD_62                 6.0             221.94              7.0   
3  1_HARDERSFIELD_63                 3.0             116.97              7.0   
4  1_HARDERSFIELD_72                 6.0             209.94              4.0   

   Ending Total $  
0          142.89  
1            0.00  
2          258.93  
3          272.93  
4          139.96

## Deliverable 4
#### Note: Because my computer could not handle the Purchases data, all answers for Deliverable 4 were calculated from the InvoicePurchases sample table using transactions with invoice dates in 2016.

### 4a

In [11]:
mostinv = invpur[(invpur["InvoiceDate"] >= "2016-01-01") & (invpur["InvoiceDate"] <= "2016-12-31")]
mostinv = mostinv.groupby(["VendorNumber","VendorName"])[["Dollars"]].sum()
mostinv = mostinv.sort_values(by = "Dollars", ascending = False)
mostinv.head(10)

Dollars
VendorNumber VendorName                              
3960         DIAGEO NORTH AMERICA INC     49733333.66
4425         MARTIGNETTI COMPANIES        26810799.70
12546        JIM BEAM BRANDS COMPANY      23539441.79
17035        PERNOD RICARD USA            23460507.21
480          BACARDI USA INC              16985053.64
1392         CONSTELLATION BRANDS INC     15187741.32
1128         BROWN-FORMAN CORP            13139921.12
9165         ULTRA BEVERAGE COMPANY LLP   12433070.97
3252         E & J GALLO WINERY           11858245.15
9552         M S WALKER INC               10294481.73

### 4b

In [12]:
mostfr = invpur[(invpur["InvoiceDate"] >= "2016-01-01") & (invpur["InvoiceDate"] <= "2016-12-31")]
mostfr = mostfr.groupby(["VendorNumber","VendorName"])[["Freight"]].sum()
mostfr = mostfr.sort_values(by = "Freight", ascending = False)
mostfr.head(10)

Freight
VendorNumber VendorName                            
3960         DIAGEO NORTH AMERICA INC     250899.75
4425         MARTIGNETTI COMPANIES        139767.62
12546        JIM BEAM BRANDS COMPANY      120230.57
17035        PERNOD RICARD USA            120130.51
480          BACARDI USA INC               86153.58
1392         CONSTELLATION BRANDS INC      77791.20
1128         BROWN-FORMAN CORP             66576.22
9165         ULTRA BEVERAGE COMPANY LLP    63888.38
3252         E & J GALLO WINERY            59594.41
9552         M S WALKER INC                52199.20

### 4c

In [13]:
abv250 = invpur[(invpur["InvoiceDate"] >= "2016-01-01") & (invpur["InvoiceDate"] <= "2016-12-31")]
abv250 = abv250.groupby(["VendorNumber","VendorName"])[["Freight", "Dollars", "Quantity"]].sum()
abv250 = abv250[(abv250["Dollars"] >= 250000)]
abv250["Freight per $"] = (abv250["Freight"] / abv250["Dollars"])
abv250["Freight per Unit"] = (abv250["Freight"] / abv250["Quantity"])
abv250 = abv250.sort_values(by = "Freight per $", ascending = False)
abv250.head(10)

Freight      Dollars  Quantity  \
VendorNumber VendorName                                                      
9625         WESTERN SPIRITS BEVERAGE CO    1746.69    323948.73     52243   
1590         DIAGEO CHATEAU ESTATE WINES    7083.48   1331574.55    183317   
9744         FREDERICK WILDMAN & SONS       3885.16    736944.55     68087   
17031        FLAG HILL WINERY & VINEYARD    1477.16    281173.13     19188   
516          BANFI PRODUCTS CORP            8034.30   1529676.79    213967   
653          STATE WINE & SPIRITS           7757.25   1481052.81    149590   
4425         MARTIGNETTI COMPANIES        139767.62  26810799.70   2544805   
8892         TRINCHERO FAMILY ESTATES      11225.72   2175075.75    340040   
8673         STE MICHELLE WINE ESTATES     15507.50   3007382.16    411345   
9815         WINE GROUP INC                26280.87   5109629.23    863638   

                                          Freight per $  Freight per Unit  
VendorNumber VendorName                                                    
9625         WESTERN SPIRITS BEVERAGE CO       0.005392          0.033434  
1590         DIAGEO CHATEAU ESTATE WINES       0.005320          0.038641  
9744         FREDERICK WILDMAN & SONS          0.005272          0.057062  
17031        FLAG HILL WINERY & VINEYARD       0.005254          0.076984  
516          BANFI PRODUCTS CORP               0.005252          0.037549  
653          STATE WINE & SPIRITS              0.005238          0.051857  
4425         MARTIGNETTI COMPANIES             0.005213          0.054923  
8892         TRINCHERO FAMILY ESTATES          0.005161          0.033013  
8673         STE MICHELLE WINE ESTATES         0.005156          0.037699  
9815         WINE GROUP INC                    0.005143          0.030430

In [14]:
abv250.tail(5)

Freight     Dollars  Quantity  \
VendorNumber VendorName                                                    
8352         LUXCO INC                     9785.35  1951124.00    271443   
3924         HEAVEN HILL DISTILLERIES     13854.16  2773519.92    346500   
3089         SIDNEY FRANK IMPORTING CO     8258.10  1656429.63    179772   
9819         TREASURY WINE ESTATES        14470.47  2903971.85    487654   
6359         OLE SMOKY DISTILLERY LLC      1876.61   377754.50    108698   

                                          Freight per $  Freight per Unit  
VendorNumber VendorName                                                    
8352         LUXCO INC                         0.005015          0.036049  
3924         HEAVEN HILL DISTILLERIES          0.004995          0.039983  
3089         SIDNEY FRANK IMPORTING CO         0.004985          0.045937  
9819         TREASURY WINE ESTATES             0.004983          0.029674  
6359         OLE SMOKY DISTILLERY LLC          0.004968          0.017264

### 4d

In [15]:
ft100 = invpur[(invpur["InvoiceDate"] >= "2016-01-01") & (invpur["InvoiceDate"] <= "2016-12-31")]
ft100 = ft100[ft100["Freight"] > 100]
ft100 = ft100[ft100["Quantity"] <= 1000]
ft100 = ft100.groupby(["VendorNumber", "VendorName"])[["Freight", "Dollars", "Quantity"]].sum()
ft100 = ft100.sort_values(by = "Freight", ascending = False)
ft100.head(10)

Freight    Dollars  Quantity
VendorNumber VendorName                                               
2561         EDRINGTON AMERICAS           2944.67  570627.52     18120
8664         STOLI GROUP,(USA) LLC         349.02   14299.98       975
7239         REMY COINTREAU USA INC        348.07    7515.38       382
653          STATE WINE & SPIRITS          303.70    5544.89       541
6785         PALM BAY INTERNATIONAL INC    242.73    5533.18       715
10000        MAJESTIC FINE WINES           218.05    7234.88       875
3924         HEAVEN HILL DISTILLERIES      200.02    7079.02       818
8352         LUXCO INC                     196.61    6139.47       819
4692         KOBRAND CORPORATION           185.93    5336.48       427
1485         CASTLE BRANDS CORP.           179.26    5420.41       320

### 4e

In [16]:
ftper = invpur[(invpur["InvoiceDate"] >= "2016-01-01") & (invpur["InvoiceDate"] <= "2016-12-31")]
ftper = ftper[ftper["Freight"] > 100]
ftper = ftper[ftper["Quantity"] <= 1000]
ftper = ftper.groupby(["VendorNumber","VendorName"])[["Freight", "Dollars", "Quantity"]].sum()
ftper["Freight per $"] = (ftper["Freight"] / ftper["Dollars"])
ftper["Freight per Unit"] = (ftper["Freight"] / ftper["Quantity"])
ftper = ftper.sort_values(by = "Freight per $", ascending = False)
ftper = ftper[["Freight per $", "Freight per Unit"]]
ftper.head(10)

Freight per $  Freight per Unit
VendorNumber VendorName                                                  
653          STATE WINE & SPIRITS              0.054771          0.561368
7239         REMY COINTREAU USA INC            0.046314          0.911178
6785         PALM BAY INTERNATIONAL INC        0.043868          0.339483
2555         DISARONNO INTERNATIONAL LLC       0.041883          0.381455
6213         MHW LTD                           0.036812          0.430825
4692         KOBRAND CORPORATION               0.034841          0.435433
1485         CASTLE BRANDS CORP.               0.033071          0.560187
8352         LUXCO INC                         0.032024          0.240061
10000        MAJESTIC FINE WINES               0.030139          0.249200
3089         SIDNEY FRANK IMPORTING CO         0.030117          0.353113

Compared to those in 4c, the vendors listed under the parameters in 4d have much higher freight per dollar ratios and freight per units ratios. For example, the highest freight per dollar ratio in 4c is 0.005392, while the highest in 4d is 0.054771.

## Deliverable 5
#### I will write code that combines the sales and purchases data to return the total contribution margin earned on each product sold in March 2016. For each product, contribution margin will be calculted by taking total sales revenue and subtracting total the total purchase prices for transactions in March 2016. I will be ignoring freight costs and excise taxes.

In [17]:
pur16 = purchases[(purchases["InvoiceDate"] >= "2016-03-01") & (purchases["InvoiceDate"] <= "2016-03-31")]
sal16 = sales[(sales["SalesDate"] >= "2016-03-01") & (sales["SalesDate"] <= "2016-03-31")]
cm = pd.merge(left = purchases, right = sal16, how = "inner", on = "InventoryId")
cm["Purchase Dollars"] = (cm["PurchasePrice"] * cm["SalesQuantity"])
cm["Contribution Margin"] = (cm["SalesDollars"] - cm["Purchase Dollars"])
cm = cm.groupby(["InventoryId"])[["Contribution Margin"]].sum()
cm = cm.sort_values(by = "Contribution Margin", ascending = False)
cm.head(10)

Contribution Margin
InventoryId                            
69_MOUNTMEND_1233              56842.56
76_DONCASTER_6570              51273.00
38_GOULCREST_3405              50450.40
34_PITMERDEN_4261              47787.74
34_PITMERDEN_1233              35691.84
67_EANVERNESS_3545             35107.60
50_MOUNTMEND_1233              34875.36
38_GOULCREST_3545              31592.80
66_EANVERNESS_3405             29488.20
76_DONCASTER_1233              28343.52